### PR TITLE
feat: eslint-plugin-node to eslint-plugin-n

### PR DESCRIPTION
**Pull request by aladdin-add**
it was forked from eslint-plugin-node v11.1.0. as the original repository seems no longer maintained.

repo link: [github.com/eslint-community/eslint-plugin-n](https://github.com/eslint-community/eslint-plugin-n)
npm link: https://www.npmjs.com/package/eslint-plugin-n
related discussion: https://github.com/mysticatea/eslint-plugin-node/issues/300

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Please provide some links/references

**Comment by aladdin-add** [comment]
yes, it's just a wip. will provide more info and sign the cla later.

**Comment by voxpelli** [comment]
I guess one challenge here is that eg the eslint rules are renamed from `node/` to `n/` and the `plugins` and/or `extends` of an eslint config also needs to reflect the new name

Can this be handled?

File: lib/config/presets/internal/replacements.ts
Comment (viceice) [inline lib/config/presets/internal/replacements.ts:630]: this should be the lowest possible version.

**Comment by viceice** [comment]
> I guess one challenge here is that eg the eslint rules are renamed from `node/` to `n/` and the `plugins` and/or `extends` of an eslint config also needs to reflect the new name
> 
> Can this be handled?

no

**Comment by voxpelli** [comment]
> no

Then sadly I think this isn't possible to do right now.

Would be nice if one could eg. show a replacement suggestion in the Dashboard issue, like:

```
### Unmaintained dependencies with community replacements:

* `eslint-plugin-node` has been superseded by `eslint-plugin-n`
```

**Comment by viceice** [comment]
renovate should simply open the replacement PR and let a maintainer fix the migration needed.

File: lib/config/presets/internal/replacements.ts
Comment (voxpelli) [inline lib/config/presets/internal/replacements.ts:629]:
```suggestion
        matchPackageNames: ['eslint-plugin-node'],
        replacementName: 'eslint-plugin-n',
```

File: lib/config/presets/internal/replacements.ts
Comment (voxpelli) [inline lib/config/presets/internal/replacements.ts:26]:
```suggestion
      'replacements:eslint-plugin-node-to-maintenance-fork',
```

File: lib/config/presets/internal/replacements.ts
Comment (voxpelli) [inline lib/config/presets/internal/replacements.ts:624]:
```suggestion
    description: 'Maintenance fork of `eslint-plugin-node`',
```

**Review by voxpelli — CHANGES_REQUESTED** [review]
The name got a bit mixed up with `npm-run-all` 😜

**Comment by voxpelli** [comment]
> renovate should simply open the replacement PR and let a maintainer fix the migration needed.

@viceice That assumes that people's tests will fail so that they realize that they need to do additional manual changes

To make matters even worse: If they are using a shared ESLint config, then it might not even be they who should make the change, it should be the upstream

File: lib/config/presets/internal/replacements.ts
Comment (aladdin-add) [reply lib/config/presets/internal/replacements.ts:629]: ops, copy-paste error 😅

File: lib/config/presets/internal/replacements.ts
Comment (aladdin-add) [reply lib/config/presets/internal/replacements.ts]: We are preparing a preview version for v17, so the latest latest is v16.

I wrote the most recent v15(`15.7.0`) here - since it supports node.js/eslint much closer to the original package, it should be easier for users to adopt. does it make sense to you?

| Version | Supported Node.js | Supported ESLint Version |
|---------|-------------------|---------------------------|
| 17.x   | `^18.18.0 \|\| ^20.9.0 \|\| >=21.1.0`  | `>=8.23.0`          |
| 16.x   | `>=16.0.0`   | `>=7.0.0`          |
| 15.x   | `>=12.22.0`  | `>=7.0.0`          |

**Comment by aladdin-add** [comment]
updated the OP, and signed the CLA.🎉

**Comment by aladdin-add** [comment]
> That assumes that people's tests will fail so that they realize that they need to do additional manual changes

Can you give an example of a situation where a ci won't fail?🤔

**Comment by voxpelli** [comment]
> > That assumes that people's tests will fail so that they realize that they need to do additional manual changes
> 
> Can you give an example of a situation where a ci won't fail?🤔

When there is no CI :) Not everyone runs their ESLint on CI

**Comment by viceice** [comment]
> > > That assumes that people's tests will fail so that they realize that they need to do additional manual changes
> > 
> > Can you give an example of a situation where a ci won't fail?🤔
> 
> When there is no CI :) Not everyone runs their ESLint on CI

that's not renovate's fault. renovate won't automerge r placements by default, so it's the developer who is responsible to check required changes before merging such PRs.

**Comment by voxpelli** [comment]
Lets try it then and see what the feedback is, would be great to get people moving to it 👍

File: lib/config/presets/internal/replacements.ts
Comment (viceice) [inline lib/config/presets/internal/replacements.ts:192]:
```suggestion
        replacementVersion: '14.0.0',
```
why not v14? it seems it's the oldest available version on npm and we tend to make the smallest step to reduce too many breaking changes.

Comment (aladdin-add) [reply lib/config/presets/internal/replacements.ts:192]: SGTM! 👍

**Review by viceice — DISMISSED** [review]
please don't force-push in future, we're doing squash merge

File: lib/config/presets/internal/replacements.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/replacements.ts:186]:
```suggestion
    description: 'Replace stale `estlint-plugin-node` with a maintained fork: `eslint-plugin-n`.',
```

How about something like this? The description should make it easy to see _what's_ changing and _why_. Also we should end sentences with a full stop (`.`). 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/presets/internal/replacements.ts
Comment (voxpelli) [inline lib/config/presets/internal/replacements.ts:186]:
```suggestion
    description: 'Replace stale `eslint-plugin-node` with a maintained fork: `eslint-plugin-n`.',
```
Typo :)

Comment (HonkingGoose) [reply lib/config/presets/internal/replacements.ts:186]: Dang it! 😄 Good catch.

**Comment by voxpelli** [comment]
@aladdin-add Can you maybe update the title of this PR to refer to `eslint-` instead of `npm-` as well? Then eg. a straight up squash commit will have the correct wording

**Comment by aladdin-add** [comment]
my bad! 😅 updated~

File: lib/config/presets/internal/replacements.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/replacements.ts:17]:
```suggestion
      'replacements:eslint-plugin-node-to-maintained-fork',
```

File: lib/config/presets/internal/replacements.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/replacements.ts:185]:
```suggestion
  'eslint-plugin-node-to-maintained-fork': {
```

**Review by HonkingGoose — COMMENTED** [review]
Small style change to the preset name maybe?

**Review by HonkingGoose — APPROVED** [review]
This approval is only for the text content. 😉 

I'll leave it to the maintainers to approve this preset, and to review if _this_ fork is the _best_ replacement.

**Comment by voxpelli** [comment]
> if _this_ fork is the _best_ replacement

Some added context here that may be helpful for them:

This is the fork that's maintained by the official [ESLint Community](https://github.com/eslint-community) – a home for high-value projects in the ESLint ecosystem where the ESLint community collectively can ensure that they continue to be maintained and not rely on any single maintainer / releaser.

**Comment by aladdin-add** [comment]
friendly ping...

**Comment by rarkins** [comment]
So I think we confirmed:
- Old package is unmaintained
- New package is legitimate

But is there a problem if this PR _breaks_ people's scripts due to migration being necessary?

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
I think we should limit current version to `^11.1.0`, so only if a user is on latest `eslint-plugin-node` major then open a replacement PR? Otherwise we suggest updates to older versions too, even if the user disabled majors.

**Review by viceice — DISMISSED** [review]

**Comment by aladdin-add** [comment]
> still conflicted

should be addressed now.

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**
